### PR TITLE
ci: run PR jobs on Ubuntu, MacOS and Windows

### PR DESCRIPTION
**Pull request by petebacondarwin**
After the cache has been primed we still get super fast jobs on Ubuntu but somewhat slower on MacOS and Windows:

Screenshot 2022-02-01 at 14 04 17

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Hmm, this is worrying. When we use the CI cache for node_modules in Windows and then run `npm install` we get this:

```
Run npm install

removed 1 package, changed 4 packages, and audited 1330 packages in 4s
```

Followed by a bunch of failures. It seems that whatever we are caching is not quite right for Windows npm... Or Windows npm is broken.

**Comment by threepointone** [comment]
I can confirm something's up with our caching strategy, but also something's up with how eslint cache even works.I faced CI failures in https://github.com/cloudflare/wrangler2/pull/348 but not locally. Had to blow away my local `.eslintcache` to get the error.

**Comment by petebacondarwin** [comment]
> Hmm, this is worrying. When we use the CI cache for node_modules in Windows and then run `npm install` we get this:
> 
> ```
> Run npm install
> 
> removed 1 package, changed 4 packages, and audited 1330 packages in 4s
> ```
> 
> Followed by a bunch of failures. It seems that whatever we are caching is not quite right for Windows npm... Or Windows npm is broken.

I worked out what the problem is here.

We were caching the `node_modules` directories but we are using npm workspaces, which create symlinks from the top level to each of the workspaces (e.g. `node_modules/wrangler` gets symlinked to `packages/wrangler`). Unfortunately, on Windows, when the cache is unpacked these symlinks are lost so npm thinks that any dependencies that had been stored locally under `packages/wrangler/node_modules` were no longer required and are removed. Hence the message above and also hence the broken checks and tests.

File: package.json [hunk 1]
```diff
@@ -33,12 +33,12 @@
   "scripts": {
     "check": "run-p check:* --aggregate-output -l",
     "check:type": "tsc -p tsconfig.json",
-    "check:lint": "eslint packages/** --cache --cache-strategy content",
-    "check:format": "prettier packages/** --check",
+    "check:lint": "eslint \"packages/**/*.[tj]s\" --cache --cache-strategy content",
```
Comment (threepointone) [inline package.json:36]: does this include tsx/jsx?

File: packages/wrangler/src/__tests__/publish.test.ts [hunk 2]
```diff
@@ -999,17 +991,14 @@ export default{
...
-      );
     });
   });
 
   describe("custom builds", () => {
     it("should run a custom build before publishing", async () => {
       writeWranglerToml({
         build: {
-          command: `echo "custom build" && echo "export default { fetch(){ return new Response(123) } }" > index.js`,
+          command: `node -e "console.log('custom build'); require('fs').writeFileSync('index.js', 'export default { fetch(){ return new Response(123) } }')"`,
```
Comment (threepointone) [inline packages/wrangler/src/__tests__/publish.test.ts:1001]: This test was specifically to also confirming that conjugated commands with `&&` ran as expected. I'm guessing you rewrote this to be x-platform compatible. Is there anything we can do to test conjugated commands, but x-platform?

**Review by threepointone — APPROVED** [review]
Looks great, very excited to see the improvements! Approving tentatively, but with a couple of questions.

File: package.json (hunk 1 above)
Comment (petebacondarwin) [reply package.json:36]: 😱  no! I don't think so....

File: packages/wrangler/src/__tests__/publish.test.ts (hunk 2 above)
Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/publish.test.ts:1001]: We can have an OS specific test. But the `&&` is not a valid operator in Windows command line.

Comment (threepointone) [reply packages/wrangler/src/__tests__/publish.test.ts:1001]: Oh well. Should be fine.

Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/publish.test.ts:1001]: I've added a test...

File: package.json (hunk 1 above)
Comment (petebacondarwin) [reply package.json]: Fixed.

**Pull request merged**